pathspider/chains/traceroute.py

```diff
@@ -0,0 +1,155 @@
+"""
+.. module:: pathspider.chains.traceroute
+   :synopsis: A flow analysis chain for traceroute messages especially icmp messages
+
+"""
+
+from pathspider.chains.base import Chain
+from pathspider.traceroute_base import INITIAL_SEQ
+import logging
+#from pathspider.base import Spider
+from straight.plugin import load
+
+chain = load("pathspider.chains", subclasses=Chain)
+
+#plug = PluggableSpider()
+#current_plugin = plug.plugin()
+
+# TODO PLUGIN works but I just take the one plugin with _trace:  ecn_trace???????????????
+
+chosen_chains = []
+
+for abc in chain:
+        if "_trace" in abc.__name__.lower():
+            chosen_chains.append(abc)
+
+#: ICMPv4 Message Type - TTL Exceeded
+ICMP4_TTLEXCEEDED = 11
+#: ICMPv6 Message Type - Time Exceeded
+ICMP6_TTLEXCEEDED = 3
+
+
+
+class tracerouteChain(Chain):
+    """
+    This flow analysis chain is the basic chain for tracerouting 
+    """
+    
+    #print(self.args)
+    
+    def new_flow(self, rec, ip):
+         """
+         For a new flow, all fields will be initialised to ``False`` or 0.
+ 
+         :param rec: the flow record
+         :type rec: dict
+         :param ip: the IP or IPv6 packet that triggered the creation of a new
+                    flow record
+         :type ip: plt.ip or plt.ip6
+         :return: Always ``True``
+         :rtype: bool
+         """
+         
+         rec['trace'] = False
+         rec['hops'] = 0
+         rec['_seq'] = 11000
+         return True
+        
+    def ip4(self, rec, ip, rev):
+        if ip.tcp and not rev: 
+            self.dest_seq(rec, ip)
+        if ip.tcp and rev:
+            self.dest_trace(rec, ip, ip.icmp)
+        return True
+            
+    def ip6(self, rec, ip6, rev):
+        if ip6.tcp and not rev: 
+            self.dest_seq(rec, ip6)
+        if ip6.tcp and rev:
+            self.dest_trace(rec, ip6, ip6.icmp6)
+        return True
+            
+    def icmp4(self, rec, ip, q, rev):
+        if rev and ip.icmp.type == ICMP4_TTLEXCEEDED:
+            try:
+                tcp_seq = ip.icmp.payload.tcp.seq_nbr  
```

Comment:
Should there be an `if ip.icmp.payload.tcp:` here? We'd want this to work for UDP too, which would just be throwing attribute errors all over the place at the moment.